django/http/request.py

```diff
@@ -176,15 +177,29 @@ def build_absolute_uri(self, location=None):
             location = '//%s' % self.get_full_path()
         bits = urlsplit(location)
         if not (bits.scheme and bits.netloc):
-            current_uri = '{scheme}://{host}{path}'.format(scheme=self.scheme,
-                                                           host=self.get_host(),
-                                                           path=self.path)
-            # Join the constructed URL with the provided location, which will
-            # allow the provided ``location`` to apply query strings to the
-            # base path as well as override the host, if it begins with //
-            location = urljoin(current_uri, location)
+            # Handle the trivial case trivially. If the location is absolute
+            # and no scheme or host (netloc) is provided we can shortcut
+            # urljoin as long as no path segments are '.' or '..'.
+            if (bits.path.startswith('/') and not bits.scheme and not bits.netloc and
```

Comment:
If we're into optimising, I believe getting the first char by index is still faster than `startswith`, which this back-of-the-napkin test seems to agree with:
```
> python3.5 -m timeit 'y = "/this/is/a/test"; y.startswith("/")'
1000000 loops, best of 3: 0.266 usec per loop
> python3.5 -m timeit 'y = "/this/is/a/test"; y[0]=="/"'
10000000 loops, best of 3: 0.0917 usec per loop
```
I don't know if its *worth* it, but there's another option.

Reply:
Good point. I'll fix that.

Reply:
Actually, not entirely sure this should be fixed here. The semantics are correct, it does read well, it's just that cpython is terrible at it.

The changes I suggest have a way bigger impact than this and if you (not necessarily you personally) don't disagree, I think readability should go first. As it stands, I hate this bit since it's hard to follow, adding that change will make it worse in that regard, for a very minor gain.

Reply:
Actually. I'll do some benchmarking tomorrow to see the effect of each of these.

Reply:
> The changes I suggest have a way bigger impact than this and if you (not necessarily you personally) don't disagree, I think readability should go first.

I don't disagree - it is Python after all :)
I merely offered it up as something additional, as it's unclear from reading the code how much difference the change *as is already implemented* is making (and the related DRF ticket only gives an overview of where time is spent as far as I can see)
If it doesn't nudge the needle of whatever benchmarks you're using to check this is faster, it's probably not worth it.

If you have any numbers to demonstrate the difference this patch brings, that'd be good to see (either on the ticket or on the PR)

Reply:
I ran this benchmark:

    timeit.timeit("request.build_absolute_uri(location='///foo/bar/')", number=1000000, globals={'request': request})
The results were:
* With my fix: 4.474267777055502
* With my fix but with bits.path[0] == '/': 4.34382488578558
* My shortcut but no cached property: 9.473239112645388 [this is to simulate running this on different requests)
* No shortcut but with cached property: 12.506602805107832 
* Original version: 17.600296460092068

There is a regression. Running without the cached property, to simulate calling it in different requests, on a path with '.':

    timeit.timeit("request.build_absolute_uri(location='/foo/./bar/')", number=1000000, globals={'request': request}))

The results:
* My version, no cached property: 19.713809736073017
* Original version: 18.129451751708984

So, the extra checks do add some overhead, but it's for an uncommon case. I don't have any evidence, obviously, but I'm fairly confident that the overwhelming uses build_absolute_url() are with an absolute path that doesn't have '.' or '..' in it and that the result is coming from reverse.

Reply:
After the suggestion by @kezabelle to replace the check for . and .. with one that doesn't use split, the new time it takes to run without the cached property for /foo/./bar/ is 18.46866210922599, bringing it within 2% of the original.